examples/recursive_translator.c

```diff
@@ -0,0 +1,79 @@
+#include <stdio.h>
+#include <ctype.h>
+#include <stdlib.h>
+
+//  Grammar:
+//      expr   -> term rest
+//      rest   -> + term {print '+'} 
+//              | - term {print '-'} 
+//              | ε
+//      term   -> 0 { print '0' }
+//              | 1 { print '1' }
+//                ...
+//              | 9 { print '9' }
+
+void get_expr();
+void term(int* lookahead);
+void match(int terminal, int* lookahead);
+
+void get_expr() {
+    int lookahead;
+
+    lookahead = getchar();
+    if (lookahead == EOF) {
+        fprintf(stderr, "error reading stdin\n");
+        exit(1);
+    }
+    term(&lookahead);
+
+    while (1) {
+        switch (lookahead) {
+            case EOF:
+                return;
+            case '+':
+                match('+', &lookahead);
+                term(&lookahead);
+                putchar('+');
+                break;
+            case '-':
+                match('-', &lookahead);
+                term(&lookahead);
+                putchar('-');
+                break;
+            default:
+                putchar('\n');
+                fprintf(stderr, "syntax error: invalid grammar\n");
+                exit(1);
+        }
```

Comment:
The original Java example probably should've errored out when there is invalid grammar (i.e. `11` is not in the grammar). `EOF` is expected for the sake of exiting the program, but everything else should be undefined in the grammar.

Might have a bootleg draft copy of the book or something.